rinja/src/html.rs

```diff
@@ -1,71 +1,131 @@
+            // SAFETY: the content of `escaped_buf` is pure ASCII
+            dest.write_str(unsafe { str::from_utf8_unchecked(&escaped_buf[..ESCAPED_BUF_LEN]) })?;
             last = index + 1;
         }
     }
-    fmt.write_str(&string[last..])
+    write_str_if_nonempty(&mut dest, &src[last..])
 }
 
 #[allow(unused)]
-pub(crate) fn write_escaped_char(mut fmt: impl fmt::Write, c: char) -> fmt::Result {
-    fmt.write_str(match (c.is_ascii(), c as u8) {
-        (true, b'"') => "&#34;",
-        (true, b'&') => "&#38;",
-        (true, b'\'') => "&#39;",
-        (true, b'<') => "&#60;",
-        (true, b'>') => "&#62;",
-        _ => return fmt.write_char(c),
-    })
+pub(crate) fn write_escaped_char(mut dest: impl fmt::Write, c: char) -> fmt::Result {
+    if !c.is_ascii() {
+        dest.write_char(c)
+    } else if let Some(escaped) = get_escaped(c as u8) {
+        let mut escaped_buf = ESCAPED_BUF_INIT;
+        [escaped_buf[2], escaped_buf[3]] = escaped;
+        // SAFETY: the content of `escaped_buf` is pure ASCII
+        dest.write_str(unsafe { str::from_utf8_unchecked(&escaped_buf[..ESCAPED_BUF_LEN]) })
+    } else {
+        // RATIONALE: `write_char(c)` gets optimized if it is known that `c.is_ascii()`
+        dest.write_char(c)
+    }
 }
 
-const MIN_CHAR: u8 = b'"';
-const MAX_CHAR: u8 = b'>';
+/// If the character needs HTML escaping, then return the decimal representation of the codepoint.
+#[inline(always)]
+fn get_escaped(byte: u8) -> Option<[u8; 2]> {
+    match byte {
+        MIN_CHAR..=MAX_CHAR => match TABLE.lookup[(byte - MIN_CHAR) as usize] {
+            0 => None,
+            escaped => Some(escaped.to_ne_bytes()),
+        },
+        _ => None,
+    }
+}
+
+#[inline(always)]
+fn write_str_if_nonempty(output: &mut impl fmt::Write, input: &str) -> fmt::Result {
+    if !input.is_empty() {
+        output.write_str(input)
+    } else {
+        Ok(())
+    }
+}
+
+/// List of characters that need HTML escaping, not necessarily in ordinal order.
+const CHARS: &[u8] = br#""&'<>"#;
+
+/// The character with the lowest codepoint that needs HTML escaping.
+const MIN_CHAR: u8 = {
+    let mut v = u8::MAX;
+    let mut i = 0;
+    while i < CHARS.len() {
+        if v > CHARS[i] {
+            v = CHARS[i];
+        }
+        i += 1;
+    }
+    v
+};
+
+/// The character with the highest codepoint that needs HTML escaping.
+const MAX_CHAR: u8 = {
+    let mut v = u8::MIN;
+    let mut i = 0;
+    while i < CHARS.len() {
+        if v < CHARS[i] {
+            v = CHARS[i];
+        }
+        i += 1;
+    }
+    v
+};
+
+/// Number of codepoints between the lowest and highest character that needs escaping, incl.
+const CHAR_RANGE: usize = (MAX_CHAR - MIN_CHAR + 1) as usize;
 
 struct Table {
     _align: [usize; 0],
-    lookup: [Option<[NonZeroU8; 2]>; (MAX_CHAR - MIN_CHAR + 1) as usize],
+    lookup: [u16; CHAR_RANGE],
 }
 
+/// For characters that need HTML escaping, the codepoint formatted as decimal digits,
```

Comment:
I think a word is missing here:

```suggestion
/// For characters that need HTML escaping, the codepoint is formatted as decimal digits,
```

Reply:
Stupid question: How can I add a "suggested change" when I review code?

Reply:
Welcome to "github trick of the day" :D

Like this:

``````
```suggestion
yo
```
``````

Reply:
Ahhh, thank you! :D